dnf/package.py

```diff
@@ -153,6 +160,22 @@ def debugsource_name(self):
         src_name = self.source_name if self.source_name is not None else self.name
         return src_name + self.DEBUGSOURCE_SUFFIX
 
+    def get_header(self):
+        """
+        Returns the rpm header of the package if it is installed. If not
+        installed, returns None. The header is not cached and retrieved from
+        rpmdb on every call. In case of a failure (e.g. when the rpmdb changes
+        between loading the data and calling this method), raises an instance
+        of PackageNotFoundError.
+        """
+        if self.repoid != "@System":
+            return None
+
+        try:
+            return next(rpm.TransactionSet().dbMatch(0, self.rpmdbid))
+        except StopIteration:
```

Comment:
It could also return `pyrpmError`. See `PyErr_SetString(pyrpmError, "rpmdb open failed");`

Reply:
I believe after switching to use `base._ts`, the transaction should be always open here, though `dbMatch` can do the opening too and throw the exception. If we should be handling that exception (it seems we don't do that anywhere), it should be in one place where the `TranscationSet` is created and not in each place the transaction is queried.

Reply:
Yup, the transaction set is the main handle to rpm, it's something you'd want to have around pretty much from start to finish, available from a central location such as base. Note though that having a transaction set does not equal having database open, the latter is still lazily opened on first use unless you explicitly do it. 

My recommendation would be explicitly opening and closing as needed for proper control. Such as knowing the database mode: if you're only querying, you can open read-only but otherwise open read-write rather than rely rpm to upgrade it at the last minute. Among other things, that would give proper locking and exact error reporting too, with a deeply combined construction like this you don't really know what it is that threw that exception. Etc.

Reply:
Oh and BTW, I realize that all is largely beyond the scope of this PR, just taking the opportunity to explain how I think this all should be handled from rpm POV.

Reply:
It's good to know, though, thanks @pmatilai. On that matter, looking at the rpm code, the error doesn't give any details on what it actually failed. It should ideally give some error description.

So it would appear the error can happen here, but the opening should really be done somewhere in base in a centralized manner. Right now I'm consistent with the rest of the code by not handling the exception :slightly_smiling_face: (the result of which is the user getting a traceback, or, in this case, the API user getting the rpm exception itself).

Reply:
Indeed, the gotcha is that when the lazy database open occurs on iterator initialization like this, rpm itself has no means of reporting the actual cause anywhere...